opm/autodiff/BlackoilModel.hpp

```diff
@@ -0,0 +1,435 @@
+/*
+  Copyright 2013, 2015 SINTEF ICT, Applied Mathematics.
+  Copyright 2015 Statoil ASA.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_BLACKOILMODEL_HEADER_INCLUDED
+#define OPM_BLACKOILMODEL_HEADER_INCLUDED
+
+#include <cassert>
+
+#include <opm/autodiff/AutoDiffBlock.hpp>
+#include <opm/autodiff/AutoDiffHelpers.hpp>
+#include <opm/autodiff/BlackoilPropsAdInterface.hpp>
+#include <opm/autodiff/LinearisedBlackoilResidual.hpp>
+#include <opm/autodiff/NewtonIterationBlackoilInterface.hpp>
+
+#include <array>
+
+struct UnstructuredGrid;
+struct Wells;
+
+namespace Opm {
+
+    namespace parameter { class ParameterGroup; }
+    class DerivedGeology;
+    class RockCompressibility;
+    class NewtonIterationBlackoilInterface;
+    class BlackoilState;
+    class WellStateFullyImplicitBlackoil;
+
+
+    /// A model implementation for three-phase black oil.
+    ///
+    /// The simulator is capable of handling three-phase problems
+    /// where gas can be dissolved in oil and vice versa. It
+    /// uses an industry-standard TPFA discretization with per-phase
+    /// upwind weighting of mobilities.
+    ///
+    /// It uses automatic differentiation via the class AutoDiffBlock
+    /// to simplify assembly of the jacobian matrix.
+    template<class Grid>
+    class BlackoilModel
+    {
+    public:
+        // ---------  Types and enums  ---------
+        typedef AutoDiffBlock<double> ADB;
+        typedef ADB::V V;
+        typedef ADB::M M;
+        typedef BlackoilState ReservoirState;
+        typedef WellStateFullyImplicitBlackoil WellState;
+
+        /// Model-specific solver parameters.
+        struct ModelParameter
```

Comment:
if this leads to merge conflicts for other people, I'd say that their patches should be merged first. I guess though, that this part of the change is probably not going to be the main problem in this case.